### PR TITLE
Alerting: Ability to edit alertmanager config as JSON via UI

### LINKED ISSUE #37213: Alerting: add option to edit entire alertmanager config as text
### Discussed in https://github.com/grafana/grafana/discussions/37212

<div type='discussions-op-text'>

<sup>Originally posted by **domasx2** July 26, 2021</sup>
Users testinng new alerting sometimes run into issues where they end up with multiple invalid contact points. This cannot be fixed using current UI, as it allows only editing single contact point at a time. If multiple are invalid, saving always fails validation.

It would be helpful to support and power users to be able to edit entire config as text to fix such issues quickly. Let's add such option, available only to Admin users.</div>

### PR DESCRIPTION
**Pull request by domasx2**
**What this PR does / why we need it**:

Adds ability to edit the Alertmanager config as json in a textare field in the "admin" view. This has been requested to enable users/support to fix broken config that cannot be fixed using "normal" UI. See linked issue for more details.

Extra: after deleting Cortex Alertmanager user config, GET config returns "alertamnager store object not found" error for a while before it starts returning empty config. Keep polling it and only complete the action once proper empty config is returned after the reset, for bettter ux.

https://user-images.githubusercontent.com/847684/127273369-9b9c94a2-7720-4db1-a7d4-db12cc3b3b61.mp4

**Which issue(s) this PR fixes**:

Fixes #37213

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by davidmparrott** [comment]
Is there anything in here that checks if the alertmanager config has been updated while they were making edits in the UI?

**Comment by domasx2** [comment]
> Is there anything in here that checks if the alertmanager config has been updated while they were making edits in the UI?

Yup! Before saving it fetches latest version and checks it against what was fetched when user opened the page. This is is true for all pages that edit am config in any way, not just this one

File: public/app/features/alerting/unified/utils/misc.ts [hunk 1]
```diff
@@ -47,3 +47,21 @@ export function recordToArray(record: Record<string, string>): Array<{ key: stri
 export function makeAMLink(path: string, alertManagerName?: string): string {
   return `${path}${alertManagerName ? `?${ALERTMANAGER_NAME_QUERY_KEY}=${encodeURIComponent(alertManagerName)}` : ''}`;
 }
+
+// keep retrying fn if it's error passes shouldRetry(error) and timeout has not elapsed yet
+export function retryWhile<T, E = Error>(
```
Comment (nathanrodman) [inline public/app/features/alerting/unified/utils/misc.ts:52]: Very slick 😎

**Review by nathanrodman — APPROVED** [review]
Nice work 👍

**Pull request merged**